syntax = "proto3";

import "google/protobuf/empty.proto";

// Services
service OrganizationService {
    rpc GetOrganizationsAsList(google.protobuf.Empty) returns (OrganizationResponseList);
    rpc GetOrganizationsAsStream(google.protobuf.Empty) returns (stream OrganizationResponse);
    rpc GetOrganizationById(GetOrganizationByIdRequest) returns (OrganizationResponse);
    rpc CreateOrganization(CreateOrganizationRequest) returns (OrganizationResponse);
    rpc UpdateOrganization(UpdateOrganizationRequest) returns (OrganizationResponse);
    rpc DeleteOrganization(DeleteOrganizationRequest) returns (google.protobuf.Empty);
    rpc FindOrganizationAsStream(FindOrganizationRequest) returns (stream OrganizationResponse);
    rpc FindOrganizationAsList(FindOrganizationRequest) returns (OrganizationResponseList);
}

// Message types
message OrganizationResponse {
    int32 id = 1;
    string name = 2;
    string description = 3;
    string country = 4;
    string industry = 5;
    string website = 6;
    int32 founded_year = 7;
    int32 employees_count = 8;
}

message OrganizationResponseList {
    repeated OrganizationResponse organizations = 1;
}

message GetOrganizationByIdRequest {
    int32 id = 1;
}

message DeleteOrganizationRequest {
    int32 id = 1;
}

message CreateOrganizationRequest {
    string name = 1;
    string description = 2;
    string country = 3;
    string industry = 4;
    string website = 5;
    int32 founded_year = 6;
    int32 employees_count = 7;
}

message UpdateOrganizationRequest {
    int32 id = 1;
    optional string name = 2;
    optional string description = 3;
    optional string country = 4;
    optional string industry = 5;
    optional string website = 6;
    optional int32 founded_year = 7;
    optional int32 employees_count = 8;
}

message FindOrganizationRequest {
    optional string name = 1;
    optional string description = 2;
    optional string country = 3;
    optional string industry = 4;
    optional string website = 5;
}
